Limit number of stages in a query

haozhun opened the pull request:
Fixes #10867.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java:
```diff
@@ -816,4 +817,16 @@ public FeaturesConfig setMaxGroupingSets(int maxGroupingSets)
+    }
+
+    @Config("max-stage-count")
+    public FeaturesConfig setMaxStageCount(int maxStageCount)
```
There’s probably a better place for this config. The name should also be prefixed with “query.” or something. Take a look at other configs.

martint replied:
See QueryManagerConfig

nezihyigitbasi replied:
Also please specify a min/max.

nezihyigitbasi commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java:
```diff
@@ -82,11 +98,24 @@ public static SubPlan createSubPlans(Session session, Metadata metadata, NodePar
+            throw new PrestoException(QUERY_TOO_MANY_STAGES, format(
+                    "Number of stages in the query (%s) exceeds the allowed maximum (%s). " +
+                            "If the query contains multiple DISTINCTs, please set the use_mark_distinct session property to false. " +
+                            "If the query contains multiple CTEs that are referenced more than once, please create temporary table for one or more of the CTEs.",
```
`create temporary table` -> `create temporary table(s)` ?

haozhun commented:
@nezihyigitbasi, please take another look.

* I added `@Min`. I didn't add a `@Max` because I don't know what it should be. If you have an idea, I can add it.
* I added session property. I put it in a separate commit, and kept the code that would allow it to just be a config property around.

nezihyigitbasi commented on presto-spi/src/main/java/com/facebook/presto/spi/StandardErrorCode.java:
```diff
@@ -60,6 +60,7 @@
     EXCEEDED_FUNCTION_MEMORY_LIMIT(0x0000_0025, USER_ERROR),
     ADMINISTRATIVELY_KILLED(0x0000_0026, USER_ERROR),
     INVALID_COLUMN_PROPERTY(0x0000_0027, USER_ERROR),
+    QUERY_TOO_MANY_STAGES(0x0000_0028, USER_ERROR),
```
How about we name this as `MAX_STAGE_COUNT_EXCEEDED`? This will make this error code sound more consistent with the other error codes, and also will make it consistent with the config/session property name.

haozhun replied:
I see
 
```
QUERY_TEXT_TOO_LARGE
PAGE_TOO_LARGE
TOO_MANY_REQUESTS_FAILED
```

and
```
EXCEEDED_FUNCTION_MEMORY_LIMIT
EXCEEDED_MEMORY_LIMIT
EXCEEDED_TIME_LIMIT
EXCEEDED_CPU_LIMIT
EXCEEDED_SPILL_LIMIT
```

I can do either `QUERY_TOO_MANY_STAGES` or `EXCEEDED_STAGE_COUNT_LIMIT`. I did the former because I find this one more similar in nature to `PAGE_TOO_LARGE`. But I don't feel strongly.

nezihyigitbasi commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java:
```diff
@@ -98,12 +94,12 @@ public SubPlan createSubPlans(Session session, Metadata metadata, NodePartitioni
 
         checkState(!isForceSingleNodeOutput(session) || subPlan.getFragment().getPartitioning().isSingleNode(), "Root of PlanFragment is not single node");
-        sanityCheckFragmentedPlan(subPlan);
+        sanityCheckFragmentedPlan(subPlan, SystemSessionProperties.getQueryMaxStageCount(session));
```
static import
